user-data-persistence/code/lib/remoteDB.js

```diff
@@ -0,0 +1,134 @@
+/**  
+ * Remote DB CRUD Interface for a Bixby UserData Structure
+ * 
+ * In Bixby, the UserData Structure looks like this:
+ * {
+ *   $id: <dbUserId>
+ *   $type: <bixbyConceptType>
+ *   <property1>: <bixbyConceptValue1>
+ *   <property2>: <bixbyConceptValue2>
+ *   ...
+ * }
+ * 
+ * In the remote DB Collection, a UserData Document looks like this:
+ * {
+ *   _id: <dbUserId>
+ *   userIdField: <bixbyUserId>
+ *   userDataField: {
+ *     <property1>: <bixbyConceptValue1>
+ *     <property2>: <bixbyConceptValue2>
+ *     ...
+ *   }
+ * }
+ * 
+ **/
+var http = require('http')
+var properties = require('./properties.js')
+
+module.exports = {
+  deleteUserData: deleteUserData, // DELETE UserData
+  getUserData: getUserData,       // READ UserData if exists, otherwise return nothing
+  putUserData: putUserData        // UPDATE UserData if exists, otherwise CREATE UserData
+}
+
+function createUserData(bixbyUserId, userData) {
+  const url = properties.get("config", "baseUrl") + properties.get("config", "collection")
+  const query = {
+    apikey: properties.get("secret", "apiKey")
+  }
+  const body = {}
+  body[properties.get("config", "userIdField")] = bixbyUserId
+  body[properties.get("config", "userDataField")] = JSON.stringify(userData)
+  const options = {
+    format: "json",
+    query: query,
+    cacheTime: 0
+  }
+  const response = http.postUrl(url, body, options)
+  if (response) {
+    userData = response[properties.get("config", "userDataField")]
+    userData.$id = response["_id"]
+    return userData
+  }
+}
+
+function deleteUserData(userData) {
+  const dbUserId = userData.$id
+  if (dbUserId) {
+    // Exists. Delete
+    const url = properties.get("config", "baseUrl") + properties.get("config", "collection") + "/" + dbUserId
+    const query = {
+      apikey: properties.get("secret", "apiKey")
+    }
+    const options = {
+      format: "json",
+      query: query,
+      cacheTime: 0
+    }
+    const response = http.deleteUrl(url, {}, options)
+    if (response) {
+      return true
+    } else {
+      return false
+    }
+  } else {
+    // Doesn't exist.
+    return
+  }
+}
+
+function getUserData(bixbyUserId) {
+  const url = properties.get("config", "baseUrl") + properties.get("config", "collection")
+  const query = {
+    apikey: properties.get("secret", "apiKey"),
+    q: "{\"" + properties.get("config", "userIdField") + "\":\"" + bixbyUserId + "\"}"
+  }
+  const options = {
+    format: "json",
+    query: query,
+    cacheTime: 0
+  }
+  const response = http.getUrl(url, options)
+  if (response && response.length === 1) {
+    const userData = response[0][properties.get("config", "userDataField")]
+    userData.$id = response[0]["_id"]
+    return userData
+  } else {
+    // Doesn't exist
+    return
+  }
+}
+
+function putUserData(bixbyUserId, userData) {
+  const dbUserId = userData.$id
+  delete userData.$id
```

Comment:
What's going on here?

Reply:
Remapping per the comment at the top of the file.

Reply:
These deletes specifically are magical.

Reply:
It require a bit of a leap to understand that Bixby exposes these without any specific setup.

Reply:
even more specifically: Is there an error if you _don't_ delete these?

Reply:
so many _specific_... my favorite word for the thread.

Reply:
To get into specifics:
The $id part is self-contained in this file; we move _id to $id in getUserData, then we move $id to _id in updateUserData. I'll add a comment explaining why we prefer to store it in $id for the Bixby side.
The $type part is less intuitive since Bixby automatically adds it. I'll add a note about that too.